File: src/main/scala/survey/ZipList.scala
package survey

import lists.NonEmptyList

case class ZipList[A](previous: List[A], current: A, next: List[A])

object ZipList {
  def fromNonEmpty[A](list: NonEmptyList[A]): ZipList[A] =
    ZipList(previous = List.empty, current = list.head, next = list.tail)

  def toList[A](zipList: ZipList[A]): List[A] =
    zipList.previous ++ (zipList.current :: zipList.next)
}

case class Question(question: String, answer: Option[String])

object Survey {
  type Survey = ZipList[Question]
  def of(questions: NonEmptyList[String]): Survey =
    ZipList.fromNonEmpty(questions.map(Question(_, answer = None)))

  def answerQuestion(survey: Survey, answer: String): Survey = {
    val currentWithAnswer = survey.current.copy(answer = Some(answer))
    survey.copy(current = currentWithAnswer)
  }

  def forwards(survey: Survey): Survey = survey.next match {
    case Nil => survey // No more questions
    case nextQuestion :: remainingQuestions =>
      ZipList(
        previous = survey.previous.appended(survey.current),
        current = nextQuestion,
        next = remainingQuestions,
      )
  }

  def back(survey: Survey): Survey = survey.previous.lastOption match {
    case None => survey // No previous questions
    case Some(previousQuestion) =>
      ZipList(
        previous = survey.previous.init,
        current = previousQuestion,
        next = survey.current :: survey.next,
      )
  }

  def printAll(survey: Survey): Unit =
    ZipList
      .toList(survey)
      .foreach(question => println(s"${question.question}: ${question.answer.getOrElse("")}"))
}

object OnlyValidStates {
  private val initial =
    Survey.of(NonEmptyList("What's your name?", "What's your favourite colour?"))

  val correct = Survey.answerQuestion(initial, "Oli")

  // Answered the wrong question
  val incorrect = Survey.answerQuestion(initial, "Red")
}

object InjectiveStateFunction {
  import survey.Survey.Survey
  def toSeparated(survey: Survey): SeparatedSurvey = {
    val questions = ZipList.toList(survey)
    SeparatedSurvey(
      questions = questions.map(_.question),
      answers = questions.map(_.answer),
      current = survey.previous.size,
    )
  }
}

File: golang/survey/ziplist.go
package survey

import "fmt"

type ZipListQuestions struct {
	previous []Question
	current  Question
	next     []Question
}

type Question struct {
	question string
	answer   *string
}

type Survey = ZipListQuestions

func SurveyOf(firstQuestion string, others ...string) Survey {
	next := make([]Question, len(others))
	for i, q := range others {
		next[i] = Question{
			question: q,
			answer:   nil,
		}
	}
	return ZipListQuestions{
		previous: make([]Question, 0),
		current: Question{
			question: firstQuestion,
			answer:   nil,
		},
		next: next,
	}
}

func answerQuestionZip(survey Survey, answer string) Survey {
	survey.current.answer = &answer
	return survey
}

// The following are easier to implement in pattern matching
func forwardsZip(survey Survey) Survey {
	if len(survey.next) == 0 {
		return survey
	}
	// Move current question into the previous list
	previous := append(survey.previous, survey.current)
	return ZipListQuestions{
		previous: previous,
		current:  survey.next[0],
		next:     survey.next[1:],
	}
}

func backZip(survey Survey) Survey {
	if len(survey.previous) == 0 {
		return survey
	}
	// Move the current question into the next list
	next := append([]Question{survey.current}, survey.next...)
	return ZipListQuestions{
		previous: survey.previous[:len(survey.previous)],
		current:  survey.previous[len(survey.previous)],
		next:     next,
	}
}

func printAllZip(survey Survey) {
	for i, question := range toList(survey) {
		answer := ""
		if question.answer != nil {
			answer = *question.answer
		}
		number := i + 1
		println(fmt.Sprintf("Question %v: %s: %v", number, question.question, answer))
	}
}

var (
	// Only Valid States
	initial           = SurveyOf("What's your name?", "What's your favourite colour?")
	correct           = answerQuestionZip(initial, "Oli")
	answerIncorrectly = func() Survey {
		firstAnswered := answerQuestionZip(initial, "Red")
		secondQuestion := forwardsZip(firstAnswered)
		return answerQuestionZip(secondQuestion, "Oli")
	}
	incorrect = answerIncorrectly()
)

func toList(survey Survey) []Question {
	questions := survey.previous
	questions = append(questions, survey.current)
	questions = append(questions, survey.next...)
	return questions
}

func toSeparated(survey Survey) SeparatedSurvey {
	questions := toList(survey)
	qs := make([]string, len(questions))
	as := make([]*string, len(questions))
	for i, question := range questions {
		qs[i] = question.question
		as[i] = question.answer
	}
	return SeparatedSurvey{
		questions: qs,
		answers:   as,
		current:   len(survey.previous),
	}
}
